Add options parsing

gabrielthomasjacobs opened the pull request:
This will allow the options/flags to be passed in any order, also allows for defaults to be set.

I am somewhat inexperienced with node, so please let me know if this is a bad approach.  
I did some testing and the options/flags are correctly passed/overwritten properly.

I pulled the latest code, then from root ran prepare:app
after, I could run the test with the new settings with the following command:

npm run test:app -- -- --port 3000 -d -v -r dot -o 'mocha-junit-reporter.config.json'

As per the PR, the following flags (and aliases) are available:  

| Flag          | Alias          | Default     | 
| ----------- | ----------- | ----------- |
| --port       | -p              |  3001        |
| --entry      | -e              |  'test.html'|
| --reporter | -r              |  'spec'        |
| --reporterOptions | -o |  undefined |
| --verbose | -v |  false  |
| --debug | -d |  false |


~I had to run the command with "-- --" before passing in the args so they would be forwarded properly... Is there a better way to do this?~ See comment below.

gabrielthomasjacobs commented:
Without the "--" the cmd flags are sent to the options object (cli.js) under the "args" key as an array of strings. You can see in the second commit, that I've added some logic to map the args to the original ordering.  
 
 Basically, what this means is that if you run it from the command line, you can run it the exact same way as before this change:  
`mocha-vite-puppeteer --reporter {reporter} --reporter-options {reporterOptions}`

If you want to use the new options (use port flag) from the command line you need a single "--":
`mocha-vite-puppeteer -- -p 3003 -o {reporterOptions} -r {reporter}`

If you are running this from a npm script, you do not need the "--":
```
"scripts":{
"test:unit": "mocha-vite-puppeteer -p 3003 -r dot"
}
```

Summary:
1. This is backwards compatible, so no migration needed on update.
2. Gives more flexibility  
3. Should give us better ability going forward to consuming configurations (especially optional ones)

larsthorup commented on cli.js:
```diff
@@ -2,21 +2,42 @@
 import * as fs from 'fs';
 import * as module from 'module';
 import * as path from 'path';
+import { getopt } from 'stdio';
```
I have not used this module before, but looks like a good choice!

larsthorup commented on cli.js:
```diff
@@ -2,21 +2,42 @@
+});
+
+if(options.args) {
+  options.reporter = options.args[0];
```
I am not sure I understand why we need these? I don't think we ever supported reading "reporter" from args[0]. Is this needed for the `npx mocha-vite-puppeteer` use case or for the `test:app` use case?

gabrielthomasjacobs replied:
There is only one edge-case where options.args would be defined, which is when a user has an npm script setup like this:
`"test": "mocha-vite-puppeteer"`  
  
And then calls `npm run test --port 3005`  
Although `npm run test 3005` will result in the same behavior.  

Calling npm in such a manner results in those params being "sent" to the "test" script and not directly to mocha-vite-puppeteer. When they do eventually reach cli.js, the stdio package thinks that those are just misc args and throws them under that args key.   
  
My plan for this was to catch an end-user mistake, but we could just remove this or replace it with a console.warn instead.

gabrielthomasjacobs replied:
I've updated this to just be a warning for any user that passes malformed args

larsthorup reviewed (changes requested):
Thank you very much for this PR! Could I ask you to also:

- rebase on top of latest `main` (I have added 3 small test related commits)
- add documentation of the options in README.md
- make sure that `test:app` actually verifies some of the option parsing, e.g. by renaming `test.html` to `mocha.html` and pass that option in `package.json`?

Let me know if you any questions regarding this.

gabrielthomasjacobs commented:
Hmm.... please let me know if that rebase looks right. I haven't done too many of those... I can always just create a new branch/pr with the changes if needed....  
  
Anyway, I've added some documentation to the readme, and added a new command in the app/package.json that will test the entry flag (-e test-flags.html) and as a result, will only launch the tests that match "*.test-flags.{js,jsx}". I've added one test that will match that, which will confirm the port flag was updated.

I had to hard-code the port, because I couldn't find a way to pass the value from the package.json to within the spec.

larsthorup commented:
Looks good. The rebase looks correct. Documentation is good! The test is fine. For some reason the build is failing. I will investigate that.

larsthorup commented on test/app/package.json:
```diff
@@ -5,7 +5,8 @@
     "build": "vite build",
     "serve": "vite preview",
-    "test": "node ../../cli"
+    "test": "node ../../cli",
```
Could you combine the two tests, so:
`"test": "node ../../cli && node ../../cli -e test-flags.html -p 3006"`
Then we will be testing both default values and argument passing during CI.

larsthorup commented on package.json:
```diff
@@ -7,6 +7,7 @@
   "dependencies": {
     "mocha": "^8.3.2",
     "puppeteer": "^8.0.0",
+    "stdio": "^2.1.1",
```
Unfortunately, I cannot reproduce the "npm ci" issue when your PR is build (https://github.com/larsthorup/mocha-vite-puppeteer/pull/8/checks?check_run_id=2978430300). 

But the error indicates that you have added the "stdio" package, as reflected in "package.json", and that is not reflected in "package-lock.json". 

Could you run "npm install" locally and include the updated "package-lock.json" in this PR. Hopefully that will make the build green 🚀

larsthorup commented:
Published https://www.npmjs.com/package/mocha-vite-puppeteer/v/1.5.0

Thank you @gabrielthomasjacobs for contributing!
